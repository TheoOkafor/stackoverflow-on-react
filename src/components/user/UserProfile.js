import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import QuestionsTable from './QuestionsTable';
import getUser from '../../actions/userActions';
import TopQuestions from '../aside/TopQuestions';


class UserProfile extends Component {
  componentDidMount() {
    this.props.getUser(this.props.match.url);
  }

  render() {
    let display;
    if (this.props.user) {
      const {
        username, created,
        questions, answers,
      } = this.props.user;

      if (questions) {
        questions.map((question) => {
          const acceptedAnswer = question.answers.filter(
            answer => answer.accepted
          );
          question.hasAccepted = acceptedAnswer.length > 0;
          return question;
        });
      }
      const questionNum = questions && questions.length;
      const recentQuestions = questions
        && questions.slice(questionNum - 5, questionNum).reverse();

      const timeJoined = new Date(created).toUTCString();
      display = (
        <div className="row-container home">
          <div className="sd-margin-12 d-margin-3">
            <div className="card margin-top-30" id="profile-summary">
              <h4
                className="margin-bottom-20"><span className="profile-username">
                  {username}</span>'s Summary</h4>

              <ul className="list list-unstyled">
                <li>Joined: <b>{timeJoined}</b></li>
                <li><b>{questions ? questionNum : 0}</b> Questions</li>
                <li><b>{answers ? answers.length : 0}</b> Answers</li>
              </ul>
            </div>
          </div>

          <div
            className="sd-margin-12 d-margin-6 margin-bottom-30"
            id="main-container">
            <div
              className="row-container margin-top-30 margin-bottom-20">
              <h3>Recent Questions by
                <span
                  className="profile-username"> {username && username}
                </span>
              </h3>
              <QuestionsTable questions={recentQuestions} />
            </div>

            <div className="row-container">
              <h3>All
                <span
                  id="all-questions"> {questions && questions.length} </span>
                  Questions by
                <span
                  className="profile-username"> {username && username}</span>
              </h3>

              <QuestionsTable questions={questions} />
            </div>
          </div>

          <div className="sd-margin-12 d-margin-3">
            <div>
              <h3 className="side-title">Top Questions with Most Answers</h3>
              <TopQuestions />
            </div>
          </div>
        </div>
      );
    } else {
      display = 'User not found';
    }
    return display;
  }
}
const mapStateToProps = state => ({
  user: state.userProfile.user,
});

UserProfile.propTypes = {
  getUser: PropTypes.func.isRequired,
  user: PropTypes.object,
  match: PropTypes.object.isRequired,
};

export default connect(mapStateToProps, { getUser })(UserProfile);
