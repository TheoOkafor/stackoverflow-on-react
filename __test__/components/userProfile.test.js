import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';

import UserProfile from '../../src/components/user/UserProfile';
import store from '../../src/store';

const match = {
  url: '/users/45'
};

test('Test that the CommentDisp displays', () => {
  const component = renderer.create(
    <Provider store={store}>
      <MemoryRouter>
        <UserProfile match={match} />
      </MemoryRouter>
    </Provider>
  );
  expect(component).toMatchSnapshot();
});
