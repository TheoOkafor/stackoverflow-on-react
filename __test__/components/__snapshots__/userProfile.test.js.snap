// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Userprofile component Test that the UserProfile displays fetched details 1`] = `
<div
  className="row-container home"
>
  <div
    className="sd-margin-12 d-margin-3"
  >
    <div
      className="card margin-top-30"
      id="profile-summary"
    >
      <h4
        className="margin-bottom-20"
      >
        <span
          className="profile-username"
        >
          chanpion
        </span>
        's Summary
      </h4>
      <ul
        className="list list-unstyled"
      >
        <li>
          Joined: 
          <b>
            Wed, 22 Aug 2018 11:13:09 GMT
          </b>
        </li>
        <li>
          <b>
            1
          </b>
           Questions
        </li>
        <li>
          <b>
            1
          </b>
           Answers
        </li>
      </ul>
    </div>
  </div>
  <div
    className="sd-margin-12 d-margin-6 margin-bottom-30"
    id="main-container"
  >
    <div
      className="row-container margin-top-30 margin-bottom-20"
    >
      <h3>
        Recent Questions by
        <span
          className="profile-username"
        >
           
          chanpion
        </span>
      </h3>
      <table
        className="table"
        id="top-questions"
      >
        <tbody>
          <tr>
            <td>
              <p>
                <a
                  className="question"
                  href="/questions/3"
                  onClick={[Function]}
                >
                  Why do you do yoga?
                </a>
              </p>
            </td>
            <td>
              <p
                className="num-answer "
              >
                1
              </p>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      className="row-container"
    >
      <h3>
        All
        <span
          id="all-questions"
        >
           
          1
           
        </span>
        Questions by
        <span
          className="profile-username"
        >
           
          chanpion
        </span>
      </h3>
      <table
        className="table"
        id="top-questions"
      >
        <tbody>
          <tr>
            <td>
              <p>
                <a
                  className="question"
                  href="/questions/3"
                  onClick={[Function]}
                >
                  Why do you do yoga?
                </a>
              </p>
            </td>
            <td>
              <p
                className="num-answer "
              >
                1
              </p>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
  <div
    className="sd-margin-12 d-margin-3"
  >
    <div>
      <h3
        className="side-title"
      >
        Top Questions with Most Answers
      </h3>
      <table
        className="table"
        id="top-questions"
      >
        <tbody />
      </table>
    </div>
  </div>
</div>
`;

exports[`Userprofile component Test that the Userprofile displays 1`] = `
<div
  className="row-container home"
>
  <div
    className="sd-margin-12 d-margin-3"
  >
    <div
      className="card margin-top-30"
      id="profile-summary"
    >
      <h4
        className="margin-bottom-20"
      >
        <span
          className="profile-username"
        />
        's Summary
      </h4>
      <ul
        className="list list-unstyled"
      >
        <li>
          Joined: 
          <b>
            Invalid Date
          </b>
        </li>
        <li>
          <b>
            0
          </b>
           Questions
        </li>
        <li>
          <b>
            0
          </b>
           Answers
        </li>
      </ul>
    </div>
  </div>
  <div
    className="sd-margin-12 d-margin-6 margin-bottom-30"
    id="main-container"
  >
    <div
      className="row-container margin-top-30 margin-bottom-20"
    >
      <h3>
        Recent Questions by
        <span
          className="profile-username"
        >
           
        </span>
      </h3>
      <table
        className="table"
        id="top-questions"
      >
        <tbody />
      </table>
    </div>
    <div
      className="row-container"
    >
      <h3>
        All
        <span
          id="all-questions"
        >
           
           
        </span>
        Questions by
        <span
          className="profile-username"
        >
           
        </span>
      </h3>
      <table
        className="table"
        id="top-questions"
      >
        <tbody />
      </table>
    </div>
  </div>
  <div
    className="sd-margin-12 d-margin-3"
  >
    <div>
      <h3
        className="side-title"
      >
        Top Questions with Most Answers
      </h3>
      <table
        className="table"
        id="top-questions"
      >
        <tbody />
      </table>
    </div>
  </div>
</div>
`;

exports[`Userprofile component Test that the Userprofile does not display 1`] = `"User not found"`;
